exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/DrillAggregateExpandGroupingSetsRule.java

```diff
@@ -0,0 +1,426 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.exec.planner.logical;
+
+import com.google.common.collect.ImmutableList;
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Aggregate;
+import org.apache.calcite.rel.core.AggregateCall;
+import org.apache.calcite.rel.logical.LogicalAggregate;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.util.ImmutableBitSet;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Planner rule that expands GROUPING SETS, ROLLUP, and CUBE into a UNION ALL
+ * of multiple aggregates, each with a single grouping set.
+ *
+ * This rule converts:
+ *   SELECT a, b, SUM(c) FROM t GROUP BY GROUPING SETS ((a, b), (a), ())
+ *
+ * Into:
+ *   SELECT a, b, SUM(c), 0 AS $g FROM t GROUP BY a, b
+ *   UNION ALL
+ *   SELECT a, null, SUM(c), 1 AS $g FROM t GROUP BY a
+ *   UNION ALL
+ *   SELECT null, null, SUM(c), 3 AS $g FROM t GROUP BY ()
+ *
+ * The $g column is the grouping ID that can be used by GROUPING() and GROUPING_ID() functions.
+ * Currently, the $g column is generated internally but stripped from the final output.
+ *
+ * TODO: Implement GROUPING() and GROUPING_ID() functions by:
+ * 1. Detecting these functions in the SELECT list during expansion
+ * 2. Rewriting them to reference the $g column (e.g., GROUPING(a) becomes bit extraction from $g)
+ * 3. Preserving the $g column in the output when these functions are used
+ */
+public class DrillAggregateExpandGroupingSetsRule extends RelOptRule {
+
+  public static final DrillAggregateExpandGroupingSetsRule INSTANCE =
+      new DrillAggregateExpandGroupingSetsRule();
+
+  private DrillAggregateExpandGroupingSetsRule() {
+    super(operand(Aggregate.class, any()), DrillRelFactories.LOGICAL_BUILDER,
+        "DrillAggregateExpandGroupingSetsRule");
+  }
+
+  @Override
+  public boolean matches(RelOptRuleCall call) {
+    final Aggregate aggregate = call.rel(0);
+
+    // Only match aggregates with multiple grouping sets
+    // Also only match logical aggregates (not physical ones)
+    return aggregate.getGroupSets().size() > 1
+        && (aggregate instanceof DrillAggregateRel || aggregate instanceof LogicalAggregate);
+  }
+
+  @Override
+  public void onMatch(RelOptRuleCall call) {
```

Comment:
Consider breaking this 346-line method into smaller helper methods to improve readability, maintainability, and ease of support.